File: leetcode/algorithms/tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    def invert(self,root):
        if root==None:
            return None;
        
        temp = self.invert(root.right);
        root.right = self.invert(root.left);
        root.left=temp;
        return root;
    
    def compare(self,n1,n2):
        if n1==None and n2==None:
            return True
        if n1==None or n2==None:
            return False
        return n1.val==n2.val and self.compare(n1.right,n2.right) and self.compare(n1.left,n2.left) 
    
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        
        if root==None:
            return True
        
        self.invert(root.right);
        
        return self.compare(root.left,root.right)
        

File: leetcode/algorithms/tree/symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean areNodesSymmetric(TreeNode t1,TreeNode t2){
        return (t1==null && t2==null) || ((t1!=null && t2!=null && t1.val==t2.val) && areNodesSymmetric(t1.left,t2.right) && areNodesSymmetric(t1.right,t2.left));
    }
    public boolean isSymmetric(TreeNode root) {
        return areNodesSymmetric(root,root);
    }
}